### PR TITLE
feat(ecs): Override Task Def Artifacts

### PR DESCRIPTION
**Pull request by paragbhingre**
Add a support to override task definitions. 

Backend changes to support [#6374](https://github.com/spinnaker/spinnaker/issues/6374)

Related PRs - 
Deck - [#9092](https://github.com/spinnaker/deck/pull/9092)
ORCA - [#4117](https://github.com/spinnaker/orca/pull/4117)

### CODE REVIEW

**Comment by piradeepk** [comment]
@paragbhingre can you add screenshots and details of the different scenarios that you manually tested?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:351]: Would you be able to point to a different spot where we have split logic like this for processing spel logic? I'm a bit hesitant in having more forked logic. For instance, what if enable SPEL processing is true but the artifact doesn't actually have any expressions (in that case, the logic btwn true and false would have to function the same). 

Is there a way we can share the logic instead of having forked logic?

Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:351]: Your concern is valid, but no where in the code we have this situation where in we need to either download server group info from somewhere else or get it from pipeline JSON. In this particular scenario we have that scenario due to which we need to have this forked logic. And it is hard to figure out if SPEL processing is true but the artifact doesn't actually have any expressions.

Comment (piradeepk) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:351]: In either case, could we have the same logic apply across both scenarios, and merge the artifact with the processed artifact (for instance, have orca provide the processed artifact if it uses spel expressions)?

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/CreateServerGroupWithArtifactsSpec.java
Comment (allisaurus) [inline clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/CreateServerGroupWithArtifactsSpec.java:674]: since there's only one container definition in this test, instead of filtering it out of the request each time (and incurring the cost of iterating over the list) we should retrieve it once and then do the assertions on that variable.

File: clouddriver-ecs/src/integration/resources/testoperations/createServerGroup-spelProcessedArtifact-EC2-targetGroupMappings.json
Comment (allisaurus) [inline clouddriver-ecs/src/integration/resources/testoperations/createServerGroup-spelProcessedArtifact-EC2-targetGroupMappings.json:56]: please indent the contents of the artifact so it appears as a child of `spelProcessedTaskDefinitionArtifact` (they're on the same vertical line right now)

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:341]: this would be more readable if we evaluated the `true` condition first and extracted the logic of each block to separate functions. then we could end up with something like:
```java
RegisterTaskDefinitionRequest requestTemplate = null;
if (description.isEvaluateTaskDefinitionArtifactExpressions()) {
    requestTemplate = getSpelProcessedArtifact();
} else {
    requestTemplate = getArtifactFromFile();
}
```

**Review by allisaurus — CHANGES_REQUESTED** [review]
Thanks @paragbhingre ! I've requested a couple changes re: simplification but overall looks good.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/description/CreateServerGroupDescription.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/description/CreateServerGroupDescription.java:98]: Yes

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/CreateServerGroupWithArtifactsSpec.java
Comment (paragbhingre) [reply clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/CreateServerGroupWithArtifactsSpec.java:673]: make sense, changed.

File: clouddriver-ecs/src/integration/resources/testoperations/createServerGroup-spelProcessedArtifact-EC2-targetGroupMappings.json
Comment (paragbhingre) [reply clouddriver-ecs/src/integration/resources/testoperations/createServerGroup-spelProcessedArtifact-EC2-targetGroupMappings.json]: oops, fixed.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java]: done.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/description/CreateServerGroupDescription.java
Comment (allisaurus) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/description/CreateServerGroupDescription.java:98]: why is this a list of `Map<Object,Object>`  and not just a map of object:object?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:339]: I'm confused why `serializedSpelProcessedTaskDefArtifact` is a list - i'm not sure it needs to be, but if it does, then as written this func will loop through and convert each item in the collection but only return the last one. If we're doing this for future proofing reasons wouldn't we want to retain the results of each conversion?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java:343]: I think we can just return this line, no need for the separate `requestTemplate` variable

Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java]: Make sense. Updated.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java]: So I have changed the type of spelProcessedTaskDefArtifact from List of Maps to Map of Objects which solves this issue. The code has been updated.

**Comment by allisaurus** [comment]
@paragbhingre changes look good, though it appears to be missing [unit tests](https://github.com/spinnaker/clouddriver/blob/master/clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy)? We should add ones at least for the happy case and the null processed artifact error case.

**Comment by paragbhingre** [comment]
@allisaurus  I have added the the unit test for Spel processed artifact only as we already do have the unit test that goes to the other side of the forked logic and test the code flow.

File: clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy
Comment (allisaurus) [inline clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy:1864]: what happens if we don't mock this? the resolved artifact is only used when we need to download the file, so running this test w/o would better ensure we're actually using the spelProcessedTaskDefinitionArtifact

File: clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy
Comment (allisaurus) [inline clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy:1869]: I know it's small/simple, but can we name this something more descriptive ("spelArtifactMap", etc.)

Comment (allisaurus) [reply clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperationSpec.groovy:1869]: also: maybe we can put a different execution role in here and then assert on it below to make sure the valued from this object is what's being used.

**Review by allisaurus — APPROVED** [review]
🔥 🔥 🔥  Approved! Soooooo stoked for this feature @paragbhingre ! Nice work! 

Keeping the 'do not merge' label on for now to ensure the Orca changes get approved/settled, but should be good to go after that (and after squashing your commits :P )

**Pull request merged**
